contracts/optimize/SlotHelper.sol

```diff
@@ -0,0 +1,159 @@
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+// 存储方法
+library SlotHelper{
+    uint internal constant RIGHTSLOTDATASHIFT = 32;
+    uint internal constant LENOFFSET = 224;
+    uint internal constant FIRSTSLOTDATASIZE = 28;
+    
+
+    function putRaw(bytes32 key , bytes memory data)
+        internal 
+        returns(bytes32 mdata)
+    {
+        // warn: if data as ptr to move,your should keep "data.length" by another ptr
+        uint len = data.length;
+        mdata = encodeMetadata(data);
+        if (len > FIRSTSLOTDATASIZE){
+            for (uint index = 0 ; index * 32 < len - FIRSTSLOTDATASIZE ; index ++){
+                assembly{
+                    data := add(data,0x20)
+                    mstore(0,add(key,index))
```

Comment:
We should not use offset 0 to store memory.   Instead, we could use a local variable.  Further, I feel we can minimize the use of assembly code here, and it can be optimized as

```
function putRaw(mapping (uint256 => bytes32) slots, bytes memory data)) {
   ...
   if (len > FIRST_SLOT_DATA_SIZE) {
      for (uint256 i = 0; i < (len - FIRST_SLOT_DATA_SIZE + 32 - 1) / 32; i ++) {
         uint256 data;
         uint256 ptr;
         assembly {
            ptr := add(data, 0x20 + FIRST_SLOT_DATA_SIZE)
            ptr := add(data, i * 32)
            data = mload(ptr)
         }
         
         slots[i] = data;
      } 
   }
   ...
}
```

Reply:
https://github.com/QuarkChain/evm-large-storage/pull/1/commits/767e1a045a004ddbede8d42736210cf46da4c3fc